include/evmc/mocked_host.hpp

```diff
@@ -0,0 +1,320 @@
+            else
+                status = EVMC_STORAGE_MODIFIED_AGAIN;
+        }
+
+        old.value = value;
+        return status;
+    }
+
+    /// Get the account's balance (EVMC Host method).
+    uint256be get_balance(const address& addr) noexcept override
+    {
+        record_account_access(addr);
+        const auto it = accounts.find(addr);
+        if (it == accounts.end())
+            return {};
+
+        return it->second.balance;
+    }
+
+    /// Get the account's code size (EVMC host method).
+    size_t get_code_size(const address& addr) noexcept override
+    {
+        record_account_access(addr);
+        const auto it = accounts.find(addr);
+        if (it == accounts.end())
+            return 0;
+        return it->second.code.size();
+    }
+
+    /// Get the account's code hash (EVMC host method).
+    bytes32 get_code_hash(const address& addr) noexcept override
+    {
+        record_account_access(addr);
+        const auto it = accounts.find(addr);
+        if (it == accounts.end())
+            return {};
+        return it->second.codehash;
+    }
+
+    /// Copy the account's code to the given buffer (EVMC host method).
+    size_t copy_code(const address& addr,
+                     size_t code_offset,
+                     uint8_t* buffer_data,
+                     size_t buffer_size) noexcept override
+    {
+        record_account_access(addr);
+        const auto it = accounts.find(addr);
+        if (it == accounts.end())
+            return 0;
+
+        const auto& code = it->second.code;
+
+        if (code_offset >= code.size())
+            return 0;
+
+        const auto n = std::min(buffer_size, code.size() - code_offset);
+
+        if (n > 0)
+            std::copy_n(&code[code_offset], n, buffer_data);
+        return n;
+    }
+
+    /// Selfdestruct the account (EVMC host method).
+    void selfdestruct(const address& addr, const address& beneficiary) noexcept override
+    {
+        record_account_access(addr);
+        recorded_selfdestructs.push_back({addr, beneficiary});
+    }
+
+    /// Call/create other contract (EVMC host method).
+    result call(const evmc_message& msg) noexcept override
+    {
+        record_account_access(msg.destination);
+
+        if (recorded_calls.empty())
+        {
+            recorded_calls.reserve(max_recorded_calls);
+            m_recorded_calls_inputs.reserve(max_recorded_calls);  // Iterators will not invalidate.
+        }
+
+        if (recorded_calls.size() < max_recorded_calls)
+        {
+            recorded_calls.emplace_back(msg);
+            auto& call_msg = recorded_calls.back();
+            if (call_msg.input_size > 0)
+            {
+                m_recorded_calls_inputs.emplace_back(call_msg.input_data, call_msg.input_size);
+                const auto& input_copy = m_recorded_calls_inputs.back();
+                call_msg.input_data = input_copy.data();
+            }
+        }
+        return result{call_result};
```

Comment:
If there is a single `call_result` it is not possible to deal with precompiles? Well, this function can be overriden. Have a look at the precompile stuff soltest is doing.

Reply:
This PR is for moving MockedHost from evmone. I will handle solidity later.